File: sample/sample-java/src/test/java/com/quantipixels/ogiri/samples/java/TokenServiceIntegrationTest.java
/*
 * Copyright (c) 2025 Quanti Pixels
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */
package com.quantipixels.ogiri.samples.java;

import static com.quantipixels.ogiri.security.core.AuthHeaderKt.ACCESS_TOKEN;
import static com.quantipixels.ogiri.security.core.AuthHeaderKt.CLIENT;
import static com.quantipixels.ogiri.security.core.AuthHeaderKt.EXPIRY;
import static com.quantipixels.ogiri.security.core.AuthHeaderKt.UID;
import static org.junit.jupiter.api.Assertions.*;

import com.quantipixels.ogiri.samples.java.entity.SampleToken;
import com.quantipixels.ogiri.samples.java.repository.SampleTokenRepository;
import com.quantipixels.ogiri.samples.java.service.SampleTokenService;
import com.quantipixels.ogiri.security.core.SecurityServiceException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@ActiveProfiles("test")
@Transactional
class TokenServiceIntegrationTest {

  @Autowired private SampleTokenService tokenService;
  @Autowired private SampleTokenRepository tokenRepository;

  private static final Long TEST_USER_ID = 1L;
  private static final String TEST_EMAIL = "user1@example.com";
  private static final String TEST_PASSWORD = "password";

  @BeforeEach
  void setUp() {
    tokenRepository.deleteAll();
    SecurityContextHolder.clearContext();
  }

  @Test
  void createNewAuthToken_withNullClient_generatesClientAndPersistsAppToken() {
    var authHeader = tokenService.createNewAuthToken(TEST_USER_ID, null, null);

    assertNotNull(authHeader.getAccessToken());
    assertNotNull(authHeader.getClient());
    assertEquals("user1", authHeader.getUid());

    var savedToken =
        tokenRepository.findByUserIdAndClient(TEST_USER_ID, authHeader.getClient()).orElse(null);
    assertNotNull(savedToken);
    assertEquals(TEST_USER_ID, savedToken.getUserId());
    assertEquals(authHeader.getClient(), savedToken.getClient());
    assertEquals("APP", savedToken.getTokenType());
  }

  @Test
  void verifyUser_authenticatesAndAppendsAuthHeaders() {
    var request = new MockHttpServletRequest("POST", "/api/auth/login");
    request.setRemoteAddr("127.0.0.1");
    var response = new MockHttpServletResponse();

    tokenService.verifyUser(request, response, TEST_EMAIL, TEST_PASSWORD);

    var authentication = SecurityContextHolder.getContext().getAuthentication();
    assertNotNull(authentication);
    assertEquals("user1", authentication.getName());
    assertNotNull(response.getHeader(ACCESS_TOKEN));
    assertNotNull(response.getHeader(CLIENT));
    assertEquals("user1", response.getHeader(UID));
    assertNotNull(response.getHeader(EXPIRY));
  }

  @Test
  void verifyUser_rejectsInvalidCredentialsWithoutCreatingAuthContext() {
    var request = new MockHttpServletRequest("POST", "/api/auth/login");
    request.setRemoteAddr("127.0.0.1");
    var response = new MockHttpServletResponse();

    assertThrows(
        SecurityServiceException.class,
        () -> tokenService.verifyUser(request, response, TEST_EMAIL, "wrong-password"));
    assertNull(SecurityContextHolder.getContext().getAuthentication());
    assertEquals(0, tokenRepository.findByUserIdOrderByUpdatedAtDesc(TEST_USER_ID).size());
  }

  @Test
  void createNewAuthToken_rotatesTokenForSameClientWhileKeepingSinglePersistedRow() {
    var first = tokenService.createNewAuthToken(TEST_USER_ID, "web", null);
    var second = tokenService.createNewAuthToken(TEST_USER_ID, "web", null);

    assertNotEquals(first.getAccessToken(), second.getAccessToken());
    var webTokens =
        tokenRepository.findByUserIdOrderByUpdatedAtDesc(TEST_USER_ID).stream()
            .filter(token -> "web".equals(token.getClient()))
            .toList();
    assertEquals(1, webTokens.size());
  }

  @Test
  void deleteToken_removesOnlyTargetedClientTokenForSameUser() {
    tokenService.createNewAuthToken(TEST_USER_ID, "mobile", null);
    tokenService.createNewAuthToken(TEST_USER_ID, "web", null);

    tokenService.deleteToken(TEST_USER_ID, "mobile");

    assertNull(tokenRepository.findByUserIdAndClient(TEST_USER_ID, "mobile").orElse(null));
    assertNotNull(tokenRepository.findByUserIdAndClient(TEST_USER_ID, "web").orElse(null));
  }

  @Test
  void revokeClient_removesTokenForClientRepresentedByHeaders() {
    var issued = tokenService.createNewAuthToken(TEST_USER_ID, "mobile", null);
    SampleToken savedToken =
        tokenRepository.findByUserIdAndClient(TEST_USER_ID, "mobile").orElse(null);
    assertNotNull(savedToken);

    var request = new MockHttpServletRequest("POST", "/api/auth/logout");
    request.addHeader(ACCESS_TOKEN, issued.getAccessToken());
    request.addHeader(CLIENT, issued.getClient());
    request.addHeader(UID, issued.getUid());
    request.addHeader(EXPIRY, issued.getExpiry());
    var response = new MockHttpServletResponse();

    tokenService.revokeClient(TEST_USER_ID, request, response);

    assertNull(tokenRepository.findByUserIdAndClient(TEST_USER_ID, "mobile").orElse(null));
    assertEquals(issued.getAccessToken(), response.getHeader(ACCESS_TOKEN));
    assertEquals("mobile", response.getHeader(CLIENT));
  }
}

File: sample/sample-kotlin/src/test/kotlin/com/quantipixels/ogiri/samples/kotlin/TokenServiceIntegrationTest.kt
/*
 * Copyright (c) 2025 Quanti Pixels
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */
package com.quantipixels.ogiri.samples.kotlin

import com.quantipixels.ogiri.samples.kotlin.repository.SampleTokenRepository
import com.quantipixels.ogiri.samples.kotlin.service.SampleTokenService
import com.quantipixels.ogiri.security.core.ACCESS_TOKEN
import com.quantipixels.ogiri.security.core.CLIENT
import com.quantipixels.ogiri.security.core.EXPIRY
import com.quantipixels.ogiri.security.core.SecurityServiceException
import com.quantipixels.ogiri.security.core.UID
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.mock.web.MockHttpServletResponse
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@ActiveProfiles("test")
@Transactional
class TokenServiceIntegrationTest {
  @Autowired private lateinit var tokenService: SampleTokenService
  @Autowired private lateinit var tokenRepository: SampleTokenRepository

  private val testUserId = 1L
  private val testEmail = "user1@example.com"
  private val testPassword = "password"

  @BeforeEach
  fun setUp() {
    tokenRepository.deleteAll()
    SecurityContextHolder.clearContext()
  }

  @Test
  fun `createNewAuthToken with null client generates client and persists APP token`() {
    val authHeader = tokenService.createNewAuthToken(testUserId, null, null)

    assertNotNull(authHeader.accessToken)
    assertNotNull(authHeader.client)
    assertEquals("user1", authHeader.uid)

    val savedToken =
        tokenRepository.findByUserIdAndClient(testUserId, authHeader.client!!).orElse(null)
    assertNotNull(savedToken)
    assertEquals(testUserId, savedToken!!.userId)
    assertEquals(authHeader.client, savedToken.client)
    assertEquals("APP", savedToken.tokenType)
  }

  @Test
  fun `verifyUser authenticates and appends auth headers`() {
    val request =
        MockHttpServletRequest("POST", "/api/auth/login").apply { remoteAddr = "127.0.0.1" }
    val response = MockHttpServletResponse()

    tokenService.verifyUser(request, response, testEmail, testPassword)

    val authentication = SecurityContextHolder.getContext().authentication
    assertNotNull(authentication)
    assertEquals("user1", authentication.name)
    assertNotNull(response.getHeader(ACCESS_TOKEN))
    assertNotNull(response.getHeader(CLIENT))
    assertEquals("user1", response.getHeader(UID))
    assertNotNull(response.getHeader(EXPIRY))
  }

  @Test
  fun `verifyUser rejects invalid credentials without creating auth context`() {
    val request =
        MockHttpServletRequest("POST", "/api/auth/login").apply { remoteAddr = "127.0.0.1" }
    val response = MockHttpServletResponse()

    assertThrows(SecurityServiceException::class.java) {
      tokenService.verifyUser(request, response, testEmail, "wrong-password")
    }
    assertNull(SecurityContextHolder.getContext().authentication)
    assertEquals(0, tokenRepository.findByUserIdOrderByUpdatedAtDesc(testUserId).size)
  }

  @Test
  fun `createNewAuthToken rotates token for same client while keeping single persisted row`() {
    val first = tokenService.createNewAuthToken(testUserId, "web", null)
    val second = tokenService.createNewAuthToken(testUserId, "web", null)

    assertNotEquals(first.accessToken, second.accessToken)
    val webTokens =
        tokenRepository.findByUserIdOrderByUpdatedAtDesc(testUserId).filter { it.client == "web" }
    assertEquals(1, webTokens.size)
  }

  @Test
  fun `deleteToken removes only targeted client token for same user`() {
    tokenService.createNewAuthToken(testUserId, "mobile", null)
    tokenService.createNewAuthToken(testUserId, "web", null)

    tokenService.deleteToken(testUserId, "mobile")

    assertNull(tokenRepository.findByUserIdAndClient(testUserId, "mobile").orElse(null))
    assertNotNull(tokenRepository.findByUserIdAndClient(testUserId, "web").orElse(null))
  }

  @Test
  fun `revokeClient removes token for client represented by headers`() {
    val issued = tokenService.createNewAuthToken(testUserId, "mobile", null)
    assertNotNull(tokenRepository.findByUserIdAndClient(testUserId, "mobile").orElse(null))
    val accessToken = requireNotNull(issued.accessToken)
    val client = requireNotNull(issued.client)
    val uid = requireNotNull(issued.uid)
    val expiry = requireNotNull(issued.expiry)

    val request =
        MockHttpServletRequest("POST", "/api/auth/logout").apply {
          addHeader(ACCESS_TOKEN, accessToken)
          addHeader(CLIENT, client)
          addHeader(UID, uid)
          addHeader(EXPIRY, expiry)
        }
    val response = MockHttpServletResponse()

    tokenService.revokeClient(testUserId, request, response)

    assertNull(tokenRepository.findByUserIdAndClient(testUserId, "mobile").orElse(null))
    assertEquals(accessToken, response.getHeader(ACCESS_TOKEN))
    assertEquals(client, response.getHeader(CLIENT))
  }
}
